### PR TITLE
Show beatmap offset adjustments in a more visual way

### PR DESCRIPTION
**Pull request by peppy**
The graph will now show the adjustment which is to be applied. The motion should help explain what the slider is doing, if a user chooses to interact with it.

I'm not changing directionality from what we have because I believe the motion feels best when the slider is moving the graph in the same direction as it is dragged. Based on twitter polls, this can go either way.

One remaining point I'm not sure about is whether we should show the "-" sign in the tooltip for the offset adjust or not.

https://user-images.githubusercontent.com/191335/156710544-7b4e7f4e-3f4e-4c1f-b05d-7a9de1368daf.mp4

Dependencies for merge conflict avoidance:

- [x] Depends on https://github.com/ppy/osu/pull/17091
- [x] Depends on https://github.com/ppy/osu/pull/17087
- [x] Depends on https://github.com/ppy/osu/pull/17068

### CODE REVIEW

**Comment by frenzibyte** [comment]
Looking at the display again, I think the negative sign should remain in the number, since the direction is not directly associated to it, i.e. I would prefer `-50ms (hit objects appear later)` than `50ms (hit objects appear later)` unless it is reworded to `50ms later`, in which I would prefer that than `-50ms later` (not that I'm suggesting to reword it).

**Comment by LeaPhant** [comment]
As a player who has been using local offset excessively for a long time my personal interpretation would be for the slider to move the numbers (and accordingly the map), as changing the offset doesn't *really* change the timing of your hits directly. There's also reading involved for timing hits at least for me as someone who never plays with Hidden and orients their hit timing  a lot on approach circles. I made a poor edit of what I think moving the slider to the left should look like:

image

(The white line shows the original offset and yellow + numbers get moved by the slider. Notice the -50 disappearing, possibly a +60 would fade in at that point)

**Comment by peppy** [comment]
Thanks for the feedback. I considered this and rejected it due to implementation complexity and the inability to see the previous/new values. I think you'll get used to it, and probably just press the button anyway.

**Comment by bdach** [comment]
Have pushed a merge conflict resolution, please double check that it's correct (I believe it should be but it wasn't the easiest).

File: osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs [hunk 1]
```diff
@@ -79,105 +104,174 @@ private void load()
...
+
+                float height = Math.Clamp(value / maxValue, minimum_height, 1);
+
+                if (height > minimum_height)
+                    boxOriginal.ResizeHeightTo(height, duration, Easing.OutQuint);
+            }
+
+            public void UpdateOffset(float adjustment)
+            {
+                bool hasAdjustment = adjustment != value;
```
Comment (bdach) [inline osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs:253]: I would make this say

```suggestion
                bool hasAdjustment = adjustment != value && adjustment / maxValue >= minimum_height;
```

Rationale is that without that latter check, little yellow dots appear at the bottom in some columns and look a bit wrong:

2022-03-05-161459_709x213_scrot

Compare with how this looks after applying:

2022-03-05-162918_623x219_scrot

File: osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs [hunk 2]
```diff
@@ -79,105 +104,174 @@ private void load()
...
+                        Anchor = Anchor.BottomCentre,
+                        Origin = Anchor.BottomCentre,
+                        Colour = Color4.Yellow,
+                        Blending = BlendingParameters.Additive,
+                        Alpha = 0.6f,
+                        Height = 0,
+                    });
+                }
+
+                boxAdjustment.ResizeHeightTo(Math.Clamp(adjustment / maxValue, 0.05f, 1), duration, Easing.OutQuint);
```
Comment (bdach) [inline osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs:272]:
```suggestion
                boxAdjustment.ResizeHeightTo(Math.Clamp(adjustment / maxValue, minimum_height, 1), duration, Easing.OutQuint);
```

is what I'm guessing the intent here was

**Comment by bdach** [comment]
I've also pushed a revert of [8115a4b](https://github.com/ppy/osu/pull/17093/commits/8115a4bb8fd9e2d53c40b8607c7ad99f0f62e9a0), which was then cherry-picked to #17071, but it was presumably pasted onto a different merge base, so the hash commit changed and the change didn't disappear after merging `master` back here. Makes history a bit horrific to look at but not sure what else can be done...

**Review by bdach — COMMENTED** [review]
And here's a pass over the rest of this. Looks generally ok, issues pointed out are minor / non-blocking.

File: osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs:272]: Yeah, not really sure how this happened. I'm 99% sure it was using the `const` when I wrote it..

File: osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs [hunk 3]
```diff
@@ -51,24 +55,45 @@ public HitEventTimingDistributionGraph(IReadOnlyList<HitEvent> hitEvents)
...
-            int[] bins = new int[total_timing_distribution_bins];
+            bins = new int[total_timing_distribution_bins];
 
-            double binSize = Math.Ceiling(hitEvents.Max(e => Math.Abs(e.TimeOffset)) / timing_distribution_bins);
+            binSize = Math.Ceiling(hitEvents.Max(e => Math.Abs(e.TimeOffset)) / timing_distribution_bins);
 
             // Prevent div-by-0 by enforcing a minimum bin size
             binSize = Math.Max(1, binSize);
 
+            Scheduler.AddOnce(updateDisplay);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs:77]: This was the lazy was of ensuring that the update would only run once if the `UpdateOffset` before `LoadComplete`. It also spreads the work out over two frames (there is a small overhead with the population of drawables to the point I did try asyncing it but decided it wasn't work that complexity overhead).

I can refactor if you think it's better to have it run immediately, but as a display component that looks roughly okay without the population complete in the same frame I think it's not a huge deal.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs:77]: I think it's fine until it turns out that it isn't at some point in the future. Was just wondering when passing by.

File: osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs [hunk 4]
```diff
@@ -51,24 +55,45 @@ public HitEventTimingDistributionGraph(IReadOnlyList<HitEvent> hitEvents)
...
+        }
+
+        public void UpdateOffset(double hitOffset)
+        {
+            this.hitOffset = hitOffset;
+            Scheduler.AddOnce(updateDisplay);
+        }
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Screens/Ranking/Statistics/HitEventTimingDistributionGraph.cs:84]: Yeah, potentially.

**Review by bdach — DISMISSED** [review]

**Comment by peppy** [comment]
Had to re-jig some test steps due to an [actual failure](https://github.com/ppy/osu/runs/5433724228?check_suite_focus=true)

**Pull request merged**